ocaml/typing/jkind.mli

```diff
@@ -31,277 +31,422 @@
-     probably should be rethought at some point. *)
+  (** [is_void t] is [Void = default_to_value_and_get t].  In particular, it will
+      default the jkind to value if needed to make this false. *)
+  val is_void_defaulting : t -> bool
+  (* CR layouts v5: When we have proper support for void, we'll want to change
+     these three functions to default to void - it's the most efficient thing
+     when we have a choice. *)
 
-  (** Prints a violation and the thing that had an unexpected jkind
-      ([offender], which you supply an arbitrary printer for). *)
-  val report_with_offender :
-    offender:(Format.formatter -> unit) -> Format.formatter -> t -> unit
+  (** Gets the layout of a jkind; returns [None] if the layout is still unknown.
+      Never does mutation. *)
+  val get_layout : t -> Layout.Const.t option
 
-  (** Like [report_with_offender], but additionally prints that the issue is
-      that a representable jkind was expected. *)
-  val report_with_offender_sort :
-    offender:(Format.formatter -> unit) -> Format.formatter -> t -> unit
+  (** Gets the maximum modes for types of this jkind. *)
+  val get_modal_upper_bounds : t -> Mode.Alloc.Const.t
 
-  (** Simpler version of [report_with_offender] for when the thing that had an
-      unexpected jkind is available as a string. *)
-  val report_with_name : name:string -> Format.formatter -> t -> unit
+  (** Gets the maximum mode on the externality axis for types of this jkind. *)
+  val get_externality_upper_bound : t -> Externality.t
+
+  (** Computes a jkind that is the same as the input but with an updated maximum
+      mode for the externality axis *)
+  val set_externality_upper_bound : t -> Externality.t -> t
+
+  (*********************************)
+  (* pretty printing *)
+
+  val format : Format.formatter -> t -> unit
+
+  (** Format the history of this jkind: what interactions it has had and why
+      it is the jkind that it is. Might be a no-op: see [display_histories]
+      in the implementation of the [Jkind] module.
+
+      The [intro] is something like "The jkind of t is". *)
+  val format_history :
+    intro:(Format.formatter -> unit) -> Format.formatter -> t -> unit
+
+  (*********************************)
+  (* debugging *)
+
+  module Debug_printers : sig
+    val t : Format.formatter -> t -> unit
+  end
+end
+
+(* Re-export some symbols *)
+
+type sort = Type.sort
+
+module Sort = Type.Sort
+module Externality = Type.Externality
+
+(** A Jkind.t is a type representing general, higher-order kinds. Its special
+    case is a zeroth-order kind of runtime-representable types (Jkind.Type.t). *)
+type t = Types.type_expr Jkind_types.t
+
+module History : sig
+  val has_imported_history : t -> bool
+
+  val update_reason : t -> Type.History.creation_reason -> t
 end
 
 (******************************)
 (* constants *)
+(* See documentation of relevant functions in Jkind.Type for details
+   - the functions here wrap them *)
 
 module Const : sig
```

Comment:
It seems unnecessary to re-export things this way. But perhaps the problem is that the interface to constants is not really so great? Probably best to get rid of these functions from within the `Const` module within `Type`.

Reply:
Well, I'm not sure - essentially these just export a short-hand of the `Jkind.Type` version. Do you think it'd be better for accessors to always include an explicit promotion, like `Jkind.of_type_jkind Jkind.Type.Primitive.any`?

Reply:
`any` is a special case (because it needs to be split between `any` and `top`). But for the others: yes, I think I'd want an explicit conversion. That will clarify what's going on.

Reply:
I just suggested to put these back in the next PR. I think that e.g. `Jkind.Primitive.value ~reasion` is noticeably clearer than `Jkind.of_type_jkind (Jkind.Type.Primitive.value ~reason)` and you end up getting this in quite a few places. @goldfirere do you object to putting them back?

Reply:
Fine. I can accept that I was perhaps overzealous in avoiding repetition here. No objection.